### ISSUE TITLE
Conan package generation for existing libraries with complex dependencie

### ISSUE THREAD

**Issue by Prasaddiwalkar** [conan-io/conan#4586]
Hi Conan,

We built ~500 libraries in our enterprise application, We have complex dependency libraries.
I have generated conanfile.py for each library using script which adds required dependency in it.

Till now everything is fine. 

Now I want to generate packages using conan export-pkg . 
I want to do this activity using some script python/batch script. I tried to traverse through the directory where each folder contains conanfile.py for library of the folder name.

So here it picks folders randomly try to generate package. Package generation fails for some dependency as it is not yet processed in the script which is expected.

Is there any tool or utility to analyse and make sequence of the package generation out of 500 libraries so that package generation should not fail for dependency

**Comment by Johnnyxy** [comment]
Hi there,
maybe the (experimental) command [conan info](https://docs.conan.io/en/latest/reference/commands/output/info.html#info-output) could help you. With that you could get the build order of a recipe. Additionally it can output JSON that you could parse.

**Comment by jgsogo** [comment]
Wow!! 500 libraries, that`s a bunch of them, we will be very pleased to hear all the feedack you want to submit to us! 🎉 

Yes, there are several commands you can use to get the dependency order, all of them associated with the `conan info` command. Have a look at the `--build-order` argument, but also to the `--dry-build` (combined with `--only=None`).

Let us know if you are able to solve your issue.

Thanks!

**Comment by Prasaddiwalkar** [comment]
Hi @jgsogo 

I am not trying to build all the libraries, I am trying to "only generate packages" for those libraries with precise dependency.
So I just created only conanfile.py for each library.

**Comment by jgsogo** [comment]
Ah! I supposed you have the _main libraries_ identified. If those directories contain just a list of libraries that might be related to each other or not, then you will have to create the dependency list writing your own script, but it is nothing Conan can do out of the box in one single command

Nevertheless, it shouldn't be that hard. I think that the following could work: export all recipes to the cache: `conan export <path/to/conanfile.py> <reference>`, and then start building them using `conan install <reference> --build=missing`, Conan will take care of the packages that are already created so you are going to generate each package only once (but generating the packages in the cache requires building them, of course).

**Comment by Prasaddiwalkar** [comment]
@jgsogo 
Is there any way to generate packages forcefully irrespective of dependency exist or not?

**Comment by jgsogo** [comment]
As I understand it, a package is a set of assets and binaries compiled from sources, and all the required dependencies are necessary to compile and generate the binaries and hence the packages. So, from this POV, it is impossible to generate a package if dependencies do not exist.

Maybe I'm not understanding what you mean with "generating packages"

**Comment by Prasaddiwalkar** [comment]
only for conan export-pkg, while upload I will have all the dependency in place if I somehow forcefully "conan export-pkg"

**Comment by jgsogo** [comment]
No, it is not possible. At least, Conan needs the dependencies to get their ids in order to compute the package-id of the one you are exporting

**Comment by Prasaddiwalkar** [comment]
@jgsogo 
I have achieved this using Topological Sorting,

1. Get all conanfile.py from independent folder.
2. Populate Graph where each Graph vertices would have defaultdict with Key as recipe name and values as dependency (Required).

code sample as following.

```python
import importlib
import os
import re
# Python program to print topological sorting of a DAG
from collections import defaultdict


class Graph:
    def __init__(self, vertices):
        self.graph = defaultdict(list)  # dictionary containing adjacency List
        self.V = vertices  # No. of vertices

    # function to add an edge to graph
    def addEdge(self, u, v):
        self.graph[u].append(v)

    # A recursive function used by topologicalSort
    def topologicalSortUtil(self, dependencyName, visited, stack):
        # Mark the current node as visited.
        visited[dependencyName] = True

        # Recur for all the vertices adjacent to this vertex
        for dependencyNameL in list(self.graph[dependencyName]):
            if not visited[dependencyNameL]:
                self.topologicalSortUtil(dependencyNameL, visited, stack)

        # Push current vertex to stack which stores result
        stack.insert(0, dependencyName)

    # The function to do Topological Sort. It uses recursive
    # topologicalSortUtil()
    def topologicalSort(self):

        # Mark all the vertices as not visited
        visitedSolution = dict()
        for solution in self.graph:
            visitedSolution[solution] = False
            dependencyList = self.graph[solution]
            for dependentSol in dependencyList:
                visitedSolution[dependentSol] = False

        sortedStack = []
        # Call the recursive helper function to store Topological
        # Sort starting from all vertices one by one
        for rootElement in list(self.graph):
            if not visitedSolution[rootElement]:
                self.topologicalSortUtil(rootElement, visitedSolution, sortedStack)
        sortedStack.reverse()

        return sortedStack

    # Class to represent a graph
    def getSolutionName(self, expression):
        regex = r".+?(?=\/)"
        matches = re.search(regex, expression)
        return matches.group(0)

    def populateDict(self, recipePath):

        content = os.listdir(recipePath)
        for pymoduleDir in content:
            if os.path.isdir(os.path.join(recipePath, pymoduleDir)):
                className = pymoduleDir.capitalize() + "Conan"
                tcmodule = importlib.import_module("recipes." + pymoduleDir + ".conanfile")
                packageClass = getattr(tcmodule, className)
                recipeName = None
                if hasattr(packageClass, "name"):
                    recipeName = packageClass.name
                if hasattr(packageClass, "requires"):
                    requireList = packageClass.requires
                    for requireSolution in requireList:
                        solutionName = self.getSolutionName(requireSolution)
                        self.addEdge(recipeName, solutionName)
        return self.topologicalSort()

```

**Comment by memsharded** [comment]
Hi @Prasaddiwalkar 

I am closing this ticket as too old.
I think custom topological sort shouldn't be needed nowadays, and Conan has plenty of tools to do a complex generation of packages, even with complex dependencies. One typical recommendation is let a ``conan export`` export all recipes first, because order does not matter to ``conan export``, then using other conan commands like ``conan install`` or ``conan graph info`` or ``conan graph build-order`` Conan can perfectly and optimally inform of what needs to be built and in what order.

Please don't hesitate to reach out, open a new ticket if necessary if you still have questions or want to revisit this with us. Thanks!
